cdap-security/src/main/java/co/cask/cdap/security/authorization/AuthorizationEnforcementService.java

```diff
@@ -0,0 +1,154 @@
+
+  @Inject
+  public AuthorizationEnforcementService(AuthorizerInstantiator authorizerInstantiator, CConfiguration cConf) {
+    this.authorizerInstantiator = authorizerInstantiator;
+    this.authorizationEnabled = cConf.getBoolean(Constants.Security.Authorization.ENABLED);
+    this.cacheEnabled = cConf.getBoolean(Constants.Security.Authorization.CACHE_ENABLED);
+    this.cacheTtlSeconds = cConf.getInt(Constants.Security.Authorization.CACHE_TTL_SECS);
+    this.authPolicyCache = new HashMap<>();
+    ReadWriteLock lock = new ReentrantReadWriteLock();
+    this.readLock = lock.readLock();
+    this.writeLock = lock.writeLock();
+  }
+
+  @Override
+  protected Scheduler scheduler() {
+    return Scheduler.newFixedDelaySchedule(0, cacheTtlSeconds, TimeUnit.SECONDS);
+  }
+
+  @Override
+  protected void runOneIteration() throws Exception {
+    if (!cacheEnabled) {
+      return;
+    }
+    LOG.debug("Running authorization enforcement service iteration...");
+    // we want to block all access to the cache while the cache update is running
+    writeLock.lock();
+    try {
+      // TODO: Should we clear the cache first? That would clear entries for potentially stale users, for whom all
+      // privileges have been revoked. However, authorization still wouldn't be affected, because for such users, after
+      // the iteration is completed, the cache will contain an empty set.
+      for (Principal principal : authPolicyCache.keySet()) {
+        updatePrivileges(principal);
+      }
+    } finally {
+      writeLock.unlock();
+    }
+  }
+
+  /**
+   * Update cache with privileges for the specified principal. Invoked by the thread that updates the cache as well as
+   * by app fabric just before starting the program, to make sure that the privileges of the user running the program
+   * are cached when the program is run.
+   *
+   * @param principal the {@link Principal} whose privileges are to be cached.
+   */
+  public void updatePrivileges(Principal principal) throws Exception {
+    if (!cacheEnabled) {
+      return;
+    }
+    Authorizer authorizer = authorizerInstantiator.get();
+    Set<Privilege> privileges = authorizer.listPrivileges(principal);
+    authPolicyCache.put(principal, privileges);
+    LOG.trace("Updated privileges for principal {} as {}", principal, privileges);
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    super.startUp();
+    LOG.info("Starting program authorization enforcement service...");
+  }
+
+  @Override
+  public void enforce(EntityId entity, Principal principal, Action action) throws Exception {
+    if (!authorizationEnabled) {
+      return;
+    }
+    if (!cacheEnabled) {
+      authorizerInstantiator.get().enforce(entity, principal, action);
+      return;
+    }
+    readLock.lock();
+    try {
+      if (authPolicyCache.containsKey(principal)) {
+        Set<Privilege> privileges = authPolicyCache.get(principal);
+        for (Privilege privilege : privileges) {
```

Comment:
hmmm, there was a problem here, it wasn't checking for the entity. fixed now, also updated test.